### PR TITLE
Update melissa build

### PR DESCRIPTION
**Pull request by mschouler**
From the [documentation](https://melissa.gitlabpages.inria.fr/melissa/):
> Melissa is a file avoiding, fault tolerant and elastic framework, to run large scale sensitivity analysis and large scale deep surrogate training on supercomputers.

### CODE REVIEW

File: var/spack/repos/builtin/packages/melissa-api/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/melissa-api/package.py]: Please use the new `maintainers` syntax/directive here and in `py-meliisa`.  See https://spack.readthedocs.io/en/latest/packaging_guide.html#maintainers.

```suggestion
    maintainers("robcaulk", "mschouler", "raffino")
```

File: var/spack/repos/builtin/packages/melissa-api/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/melissa-api/package.py]: Please remove this comment.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: As mentioned in other note, please use the new directive syntax.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: Please remove this comment.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Confirmed the version branches.  Asking for a few minor tweaks wrt maintainers and unneeded comments.

**Comment by tldahlgren** [comment]
@robcaulk @raffino Please confirm you are willing to be maintainers for this package.  See  https://spack.readthedocs.io/en/latest/packaging_guide.html#maintainers for more information on the role.

**Comment by tldahlgren** [comment]
Adding @adamjstewart to the PR since the update includes adding a python package (`py-melissa`).

**Comment by mschouler** [comment]
> Confirmed the version branches. Asking for a few minor tweaks wrt maintainers and unneeded comments.

Thank you for your quick review, all requested changes were implemented.

File: var/spack/repos/builtin/packages/melissa-api/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/melissa-api/package.py]: You could prob delete this function if you aren't going to use it

File: var/spack/repos/builtin/packages/melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/melissa/package.py:1]: This package should first be deprecated instead of just removing it: https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: There is already a package called "melissa" on PyPI that is different than this one: https://pypi.org/project/melissa/

For naming, we try to reserve `py-<name>` for PyPI packages so that things listed in `install_requires` always match the package name. You'll have to choose a different name for this package.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]:
```suggestion
    depends_on("py-setuptools@46.4:", type=("build"))
```

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]:
```suggestion
    depends_on("py-python-rapidjson@1.9:", type="run")
```

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: Assuming that everything in requirements.txt is actually required, you'll need to add a package for plotext

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: This doesn't match the current file

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: Where do these come from?

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa/package.py]: Are these required or optional? It would be better to add variants for these that default to false.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa/package.py]: `plotext` is used to monitor the progress of a Melissa study but it is not a vital component of the framework. Since we are not involved in the development of `plotext`, we would rather not add a package for it. Instead we will make it a dev requirement.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa/package.py]: These are required.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa/package.py]: Version 1.8 is the highest available on Spack. We will bring it down in our requirements.

File: var/spack/repos/builtin/packages/py-melissa/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa/package.py]: Sorry for the confusion here.

Melissa is designed to handle two main tasks:
1. performing online sensitivity analysis which relies on what we call the SA server,
2. performing online deep surrogate training which relies on the so called DL server.

In order to make the DL server agnostic to the deep learning framework the user is interested in, it can either be used with the `torch` or the `tensorflow` child server. To alleviate the requirement management, we hence decided to only provide the following Spack variants:
- the default that corresponds to the SA server and the main requirements,
- the **DL** that corresponds to the DL server and that comes with both frameworks,
- the **TORCH** that corresponds to the DL server with `torch` only,
- and the **TF** that corresponds to the DL server with `tensorflow` only.

To make things clearer we will remove the DL variant and add comments to `requirements_deep_learning.txt` in order to make the `torch`/`tensorflow` choice more evident. I hope this will answer your concerns.

File: var/spack/repos/builtin/packages/melissa/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/melissa/package.py]: You'll also want to deprecate this one if you're trying to remove this recipe entirely in a future release

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa-core/package.py]: We usually try to keep variants lowercase

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa-core/package.py:34]:
```suggestion
    depends_on("py-mpi4py@3.1.3", type="run")
```
According to requirements.txt, only 3.1.3 is supported

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa-core/package.py:37]: I know you said you'll update this in the future, but you can always add a new version of rapidjson to Spack, it's relatively easy

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa-core/package.py]: Shouldn't this be 1.12.1?

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-melissa-core/package.py]: What about hostlist?

**Review by adamjstewart — COMMENTED** [review]
Remaining comments are only suggestions. It sounds like this package is still relatively new and the dependency management system is a WIP, so I'll let you use your best judgement and decide what the dependencies really are. Let me know if you want to leave things as is and I can merge when you're ready.

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa-core/package.py:34]: Good point, we will make it a non-strict version dependency in the requirement file.

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa-core/package.py:37]: Indeed, the [recipe](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-python-rapidjson/package.py) looks pretty straightforward. Thanks for the suggestion, I will look into it if I have the chance.

File: var/spack/repos/builtin/packages/py-melissa-core/package.py
Comment (mschouler) [reply var/spack/repos/builtin/packages/py-melissa-core/package.py]: Nice catch, thank you. We should be able to remove this dependency.

**Comment by mschouler** [comment]
> Remaining comments are only suggestions. It sounds like this package is still relatively new and the dependency management system is a WIP, so I'll let you use your best judgement and decide what the dependencies really are. Let me know if you want to leave things as is and I can merge when you're ready.

Thank you very much for your help and time. Indeed, the way we manage our python dependencies is a bit messy but building a Spack package for Melissa helped us make things cleaner. It should be good to go now.

**Review by tldahlgren — APPROVED** [review]
Re-confirmed home page and version branches/sha256.  Mainly clearing my Request changes as I'll still defer to @adamjstewart for the merge.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
